xDiffCollection/DiffCollection.swift

```diff
@@ -0,0 +1,111 @@
+    private var bins:[CollectionBin<T>] = []
+    private init(bins:[CollectionBin<T>]) {
+        self.bins = bins
+    }
+    
+    public var description: String {
+        var resp = ""
+        for bin in bins {
+            resp += "Bin\n"
+            resp += "\(bin)"
+        }
+        return resp
+    }
+    
+    public init(filters:[DiffCollectionFilter<T>]) {
+        for (idx, binFilter) in filters.enumerated() {
+            bins.append(CollectionBin(index: idx, filter: binFilter.filter, elements: []))
+        }
+    }
+    
+    public mutating func update(element:T) -> DiffCollectionResult {
+        var deletedItems : [IndexPath] = []
+        var addedItems : [IndexPath] = []
+        var updatedItems : [IndexPath] = []
+                
+        var resp : [CollectionBin<T>] = []
+        for (binIndex, bin) in bins.enumerated() {
+            if bin.filter(element) {
+                let (newBin,result) =  bin.updated(element)
+                switch result.type {
+                    case .added:
+                        addedItems.append(IndexPath(row: result.idx, section: binIndex))
+                    case .updated:
+                        updatedItems.append(IndexPath(row: result.idx, section: binIndex))
+                    default:
+                        fatalError("bin.updated action should return either added or updated statuses")
+                }
+                resp.append(newBin)
+            } else {
+                //Element doesn't belong to this bin. Delete it if found in the bin.
+                let (newBin,result) = bin.removed(element)
+                if result.idx > -1 {
+                    deletedItems.append(IndexPath(row: result.idx, section: binIndex))
+                }
+                resp.append(newBin)
+            }
+        }
+        
+        bins = resp
+        return DiffCollectionResult(deleted: deletedItems,
+                                      added:   addedItems,
+                                      updated: updatedItems)
+    }
+    
+    public mutating func delete(element:T) -> DiffCollectionResult {
+        var deletedItems : [IndexPath] = []
+        var resp : [CollectionBin<T>] = []
+        for (binIndex, bin) in bins.enumerated() {
+            let (newBin,result) = bin.removed(element)
+            if result.idx > -1 {
+                deletedItems.append(IndexPath(row: result.idx, section: binIndex))
+            }
+            resp.append(newBin)
+        }
+        bins = resp
+        return DiffCollectionResult(deleted: deletedItems,
+                                      added:   [],
+                                      updated: [])
+    }
+    
+    public func element(atIndexPath path:IndexPath) -> T? {
+        if path.section < bins.count {
+            return bins[path.section].element(at: path.row)
+        }
+        return nil
+    }
+    
+    public func numberOfElements(inSection binIndex:Int) -> Int? {
```

Comment:
why optional? Wouldn't it make sense to return 0 if we request for the bin that is not exists? It might help to facilitate life on the client side. What do you think?

Reply:
Agree!